### PR TITLE
feat(customers): Generate Checkout URL

### PR DESCRIPTION
**Pull request by jdenquin**
- We want to be able to generate a new checkout url for a defined customer
- Add `POST /api/v1/customers/:external_id/checkout_url`

### CODE REVIEW

File: app/services/customers/generate_checkout_url_service.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    def initialize(customer:)
+      @customer = customer
+      @provider_customer = customer&.provider_customer
+
+      super
+    end
+
+    def call
+      return result.not_found_failure!(resource: 'customer') if customer.blank?
+      return result.service_failure!(code: 400, message: 'no payment provider linked to this customer') if provider_customer.blank?
```
Comment (jdenquin) [inline app/services/customers/generate_checkout_url_service.rb:14]: Style/IfUnlessModifier: Modifier form of `if` makes the line too long. (https://rubystyle.guide#if-as-a-modifier)

Comment (jdenquin) [inline app/services/customers/generate_checkout_url_service.rb:14]: Layout/LineLength: Line is too long. [131/120] (https://rubystyle.guide#max-line-length)

File: app/services/customers/generate_checkout_url_service.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+      result = provider_customer.service.generate_checkout_url(send_webhook: false)
```
Comment (jdenquin) [inline app/services/customers/generate_checkout_url_service.rb:16]: Style/RedundantAssignment: Redundant assignment before returning detected.

File: app/services/payment_provider_customers/adyen_service.rb [hunk 3]
```diff
@@ -21,22 +21,22 @@ def create
...
+      result.checkout_url = res.response['url']
 
       SendWebhookJob.perform_later(
         'customer.checkout_url_generated',
         customer,
-        checkout_url:,
-      )
-
-      result.checkout_url = checkout_url
+        checkout_url: result.checkout_url,
```
Comment (jdenquin) [inline app/services/payment_provider_customers/adyen_service.rb:37]: Style/MultilineIfModifier: Favor a normal if-statement over a modifier clause in a multiline statement. (https://rubystyle.guide#no-multiline-if-modifiers)

File: app/services/payment_provider_customers/adyen_service.rb [hunk 4]
```diff
@@ -21,22 +21,22 @@ def create
...
+      ) if send_webhook
+
```
Comment (jdenquin) [inline app/services/payment_provider_customers/adyen_service.rb:39]: Layout/TrailingWhitespace: Trailing whitespace detected. (https://rubystyle.guide#no-trailing-whitespace)

File: app/services/payment_provider_customers/gocardless_service.rb [hunk 5]
```diff
@@ -25,17 +25,18 @@ def create
...
       billing_request_flow = create_billing_request_flow(billing_request.id)
 
+      result.checkout_url = billing_request_flow.authorisation_url
+
       SendWebhookJob.perform_later(
         'customer.checkout_url_generated',
         customer,
-        checkout_url: billing_request_flow.authorisation_url,
-      )
+        checkout_url: result.checkout_url,
```
Comment (jdenquin) [inline app/services/payment_provider_customers/gocardless_service.rb:37]: Style/MultilineIfModifier: Favor a normal if-statement over a modifier clause in a multiline statement. (https://rubystyle.guide#no-multiline-if-modifiers)

File: app/services/payment_provider_customers/stripe_service.rb [hunk 6]
```diff
@@ -103,12 +103,12 @@ def generate_checkout_url
       )
 
       result.checkout_url = res['url']
-
+
```
Comment (jdenquin) [inline app/services/payment_provider_customers/stripe_service.rb:106]: Layout/TrailingWhitespace: Trailing whitespace detected. (https://rubystyle.guide#no-trailing-whitespace)

File: app/services/payment_provider_customers/stripe_service.rb [hunk 7]
```diff
@@ -103,12 +103,12 @@ def generate_checkout_url
...
 
       result.checkout_url = res['url']
-
+      
       SendWebhookJob.perform_later(
         'customer.checkout_url_generated',
         customer,
         checkout_url: result.checkout_url,
-      )
+      ) if send_webhook
```
Comment (jdenquin) [inline app/services/payment_provider_customers/stripe_service.rb:111]: Style/MultilineIfModifier: Favor a normal if-statement over a modifier clause in a multiline statement. (https://rubystyle.guide#no-multiline-if-modifiers)

File: spec/services/customers/generate_checkout_url_service_spec.rb [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+        expect(result.checkout_url).to eq('http://foo.bar')
+      end
+    end
+
+    context 'when customer is blank' do
+      it 'returns an error' do
+        result = described_class.new(customer: nil).call
+
+        aggregate_failures do
+          expect(result.success?).to be_falsey
```
Comment (jdenquin) [inline spec/services/customers/generate_checkout_url_service_spec.rb:45]: RSpec/PredicateMatcher: Prefer using `be_success` matcher over `success?`. (https://rspec.rubystyle.guide/#predicate-matchers, https://www.rubydoc.info/gems/rubocop-rspec/RuboCop/Cop/RSpec/PredicateMatcher)

File: spec/services/customers/generate_checkout_url_service_spec.rb [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+        end
+      end
+    end
+
+    context 'when payment provider is blank' do
+      it 'returns an error' do
+        result = generate_checkout_url_service.call
+
+        aggregate_failures do
+          expect(result.success?).to be_falsey
```
Comment (jdenquin) [inline spec/services/customers/generate_checkout_url_service_spec.rb:56]: RSpec/PredicateMatcher: Prefer using `be_success` matcher over `success?`. (https://rspec.rubystyle.guide/#predicate-matchers, https://www.rubydoc.info/gems/rubocop-rspec/RuboCop/Cop/RSpec/PredicateMatcher)

File: app/services/payment_provider_customers/stripe_service.rb [hunk 10]
```diff
@@ -92,8 +92,8 @@ def check_payment_method(payment_method_id)
       result.single_validation_failure!(field: :payment_method_id, error_code: 'value_is_invalid')
     end
 
-    def generate_checkout_url
-      return result unless customer.organization.webhook_endpoints.any?
+    def generate_checkout_url(send_webhook: true)
+      return result unless (customer.organization.webhook_endpoints.any? || !send_webhook)
```
Comment (jdenquin) [inline app/services/payment_provider_customers/stripe_service.rb:96]: Style/ParenthesesAroundCondition: Don't use parentheses around the condition of an `unless`. (https://rubystyle.guide#no-parens-around-condition)

File: app/services/customers/generate_checkout_url_service.rb [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+    def initialize(customer:)
+      @customer = customer
+      @provider_customer = customer&.provider_customer
+
+      super
+    end
+
+    def call
+      return result.not_found_failure!(resource: 'customer') if customer.blank?
+      return result.service_failure!(
```
Comment (jdenquin) [inline app/services/customers/generate_checkout_url_service.rb:14]: Style/MultilineIfModifier: Favor a normal if-statement over a modifier clause in a multiline statement. (https://rubystyle.guide#no-multiline-if-modifiers)

File: app/services/customers/generate_checkout_url_service.rb [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+        code: 400,
+        message: 'no payment provider linked to this customer'
```
Comment (jdenquin) [inline app/services/customers/generate_checkout_url_service.rb:16]: Style/TrailingCommaInArguments: Put a comma after the last parameter of a multiline method call. (https://rubystyle.guide#no-trailing-params-comma)

File: app/controllers/api/v1/customers_controller.rb [hunk 13]
```diff
@@ -70,6 +70,24 @@ def destroy
         end
       end
 
+      def checkout_url
+        customer = current_organization.customers.find_by(external_id: params[:customer_external_id])
+
+        result = ::Customers::GenerateCheckoutUrlService.call(customer:)
+
+        if result.success?
+          render(
```
Comment (vincent-pochet) [inline app/controllers/api/v1/customers_controller.rb:79]: What about using the `V1::PaymentProviders::CustomerCheckoutSerializer`?
It's used in the webhook for the same purpose

File: app/services/customers/generate_checkout_url_service.rb [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+
+      super
+    end
+
+    def call
+      return result.not_found_failure!(resource: 'customer') if customer.blank?
+
+      if provider_customer.blank?
+        return result.service_failure!(
+          code: 400,
```
Comment (vincent-pochet) [inline app/services/customers/generate_checkout_url_service.rb:17]: Code is supposed to be a human readable code here. The controller is already responsible for translating the error class into an HTTP code. 

Other question, we usually keep service_failure for internal service error. I think what we need here is either a `precondition_failure`  (HTTP 412, not implemented yet), or a `single_validation_failure`. What do you think?

File: app/models/payment_provider_customers/base_customer.rb [hunk 15]
```diff
@@ -35,6 +35,10 @@ def provider_payment_methods
       nil
     end
 
+    def service
```
Comment (vincent-pochet) [inline app/models/payment_provider_customers/base_customer.rb:38]: Not a huge fan of this. `service` sounds really generic, and somehow breaks the single responsibility principle. What do you think of using the factory pattern (similar to the `BillableMetrics::AggregationFactory`) to initialize the right service based on the customer?

File: app/services/customers/generate_checkout_url_service.rb (hunk 14 above)
Comment (jdenquin) [reply app/services/customers/generate_checkout_url_service.rb:17]: I would prefer a `422` over a `412`

File: app/models/payment_provider_customers/base_customer.rb (hunk 15 above)
Comment (jdenquin) [reply app/models/payment_provider_customers/base_customer.rb:38]: so having a `service_class` on the model?

File: app/controllers/api/v1/customers_controller.rb [hunk 16]
```diff
@@ -70,6 +70,24 @@ def destroy
...
+            json: ::V1::PaymentProviders::CustomerCheckoutSerializer.new(
+              customer,
+              root_name: 'customer',
+              checkout_url:result.checkout_url,
```
Comment (jdenquin) [inline app/controllers/api/v1/customers_controller.rb:83]: Layout/HashAlignment: Align the keys of a hash literal if they span more than one line.

Comment (jdenquin) [inline app/controllers/api/v1/customers_controller.rb:83]: Layout/SpaceAfterColon: Space missing after colon. (https://rubystyle.guide#spaces-operators)

File: app/services/payment_provider_customers/factory.rb [hunk 17]
```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+module PaymentProviderCustomers
+  class Factory
+    def self.new_instance(provider_customer:)
+      service_class(provider_customer).new(provider_customer)
+    end
+
+    private
```
Comment (jdenquin) [inline app/services/payment_provider_customers/factory.rb:9]: Lint/UselessAccessModifier: Useless `private` access modifier.

File: app/services/payment_provider_customers/factory.rb [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+
+    def self.service_class(provider_customer)
```
Comment (jdenquin) [inline app/services/payment_provider_customers/factory.rb:11]: Lint/IneffectiveAccessModifier: `private` (on line 9) does not make singleton methods private. Use `private_class_method` or `private` inside a `class << self` block instead.

File: app/models/payment_provider_customers/base_customer.rb (hunk 15 above)
Comment (ivannovosad) [reply app/models/payment_provider_customers/base_customer.rb]: I agree with @vincent-pochet that model shouldn't know anything about the service (or a service class).

**Review by vincent-pochet — APPROVED** [review]

**Pull request merged**